src/pointer-analysis/value_set.h

```diff
@@ -211,11 +211,11 @@ class value_sett
     const exprt &expr,
     const namespacet &ns);
 
-  void apply_code(
+  virtual void apply_code(
```

Comment:
In general `public virtual` methods in non-abstract classes are a code smell, as they have two many responsibilities. They specify both the public interface, and a customization point for derived classes (which is an implementation detail). Ideally, the two responsibilities should be decoupled by providing a `private virtual` method which is called by a `public` (non-virtual) method. Then, if interface requirements change, the interface can be adjusted independently of the implementation (and vice versa).

I'm particularly concerned about retroactively making this method virtual. Pre-existing clients which always expect this method to do the same thing could be broken when derived classes change its behavior.